src/Sepes.Infrastructure/Service/PublicIpService.cs

```diff
@@ -0,0 +1,132 @@
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Sepes.Infrastructure.Service
+{
+    public class PublicIpService : IPublicIpService
+    {
+        readonly IConfiguration _configuration;
+        readonly ILogger _logger;
+
+        string _serverPublicIp;
+
+        readonly SemaphoreSlim _semaphore = new SemaphoreSlim(1);
+
+        public PublicIpService(IConfiguration configuration, ILogger<PublicIpService> logger)
+        {
+            _configuration = configuration;
+            _logger = logger;
+        }
+
+        public async Task<string> GetServerPublicIp()
+        {
+            try
+            {
+                if (String.IsNullOrWhiteSpace(_serverPublicIp))
+                {
+                    await _semaphore.WaitAsync();
+                    _serverPublicIp = await GetServerPublicIpFromExternalService();
+                }
+
+                return _serverPublicIp;
+            }           
+            finally
+            {
+                _semaphore.Release();
+            }
+        }
+
+        async Task<string> GetServerPublicIpFromExternalService()
+        {
+            var tryCount = 1;
+
+            try
+            {
+                while (tryCount < 3)
+                {
+                    using (var client = new HttpClient())
+                    {
+                        var ipUrls = ConfigUtil.GetCommaSeparatedConfigValueAndThrowIfEmpty(_configuration, ConfigConstants.SERVER_PUBLIC_IP_URLS);
+
+                        var errorMessageSb = new StringBuilder();
+
+                        foreach (var curIpUrl in ipUrls)
+                        {
+                            HttpResponseMessage responseMessage = null;
+
+                            try
+                            {
+                                responseMessage = await client.GetAsync(curIpUrl);
+
+                                if (responseMessage.IsSuccessStatusCode)
+                                {
+                                    var responseString = await responseMessage.Content.ReadAsStringAsync();
+
+                                    var deserializedResponse = JsonConvert.DeserializeObject<IpAddressResponse>(await responseMessage.Content.ReadAsStringAsync());
+
+                                    return deserializedResponse.ip;
+                                }
+                                else
+                                {
+                                    errorMessageSb.AppendLine(await GetPublicIpErrorMessage(curIpUrl, responseMessage: responseMessage));
+                                }
+                            }
+                            catch (Exception ex)
+                            {
+                                errorMessageSb.AppendLine(await GetPublicIpErrorMessage(curIpUrl, responseMessage: responseMessage, exception: ex));
+                            }
+                        }                      
+                    }
+                }
+
```

Comment:
Is it more correct to use a Do while loop? It will run minimum once time anyways. Does not matter really though